plutus-core/cost-model/budgeting-bench/Benchmarks/Values.hs

```diff
@@ -0,0 +1,343 @@
+
+lookupCoinArgs :: StdGen -> [(ByteString, ByteString, Value)]
+lookupCoinArgs gen = runStateGen_ gen \(g :: g) -> do
+  {- Exhaustive power-of-2 combinations for BST worst-case benchmarking.
+
+     Tests all combinations of sizes from powers and half-powers of 2.
+     For each integer n ∈ {1..10}, includes both 2^n and 2^(n+0.5) ≈ 2^n * √2.
+
+     This provides:
+     - 400 total test points (20 × 20 grid)
+     - Complete systematic coverage of depth range 2 to 21
+     - Finer granularity between powers of 2
+     - All distribution patterns (balanced, outer-heavy, inner-heavy)
+
+     Size values: 2, 3, 4, 6, 8, 11, 16, 23, 32, 45, 64, 91, 128, 181,
+                  256, 362, 512, 724, 1024, 1448
+
+     Depth coverage:
+     - Minimum depth: log₂(2) + log₂(2) ≈ 2
+     - Maximum depth: log₂(1448) + log₂(1448) ≈ 21
+  -}
+  let
+    -- Generate powers of 2 and their geometric means (half-powers)
+    sizes =
+      [ 2 ^ n -- 2^n
+      | n <- [1 .. 10 :: Int]
+      ]
+        ++ [ round @Double (2 ^ n * sqrt 2) -- 2^(n+0.5)
+           | n <- [1 .. 10 :: Int]
+           ]
+
+  sequence
+    -- Generate worst-case lookups for each size combination
+    [ withWorstCaseSearchKeys (generateConstrainedValueWithMaxPolicy numPolicies tokensPerPolicy g)
+    | numPolicies <- sizes
+    , tokensPerPolicy <- sizes
+    ]
+
+-- | Add worst-case search keys targeting the max-size inner map
+withWorstCaseSearchKeys :: (Monad m) => m (Value, K, K) -> m (ByteString, ByteString, Value)
+withWorstCaseSearchKeys genValueWithKeys = do
+  (value, maxPolicyId, deepestToken) <- genValueWithKeys
+  pure (Value.unK maxPolicyId, Value.unK deepestToken, value)
+
+----------------------------------------------------------------------------------------------------
+-- ValueContains -----------------------------------------------------------------------------------
+
+valueContainsBenchmark :: StdGen -> Benchmark
+valueContainsBenchmark gen =
+  createTwoTermBuiltinBenchElementwiseWithWrappers
+    (ValueLogOuterSizeAddLogMaxInnerSize, ValueTotalSize)
+    -- Container: sum of log sizes, Contained: totalSize
+    ValueContains -- the builtin fun
+    [] -- no type arguments needed (monomorphic builtin)
+    (valueContainsArgs gen) -- the argument combos to generate benchmarks for
+
+valueContainsArgs :: StdGen -> [(Value, Value)]
+valueContainsArgs gen = runStateGen_ gen \g -> do
+  {- ValueContains performs multiple LookupCoin operations (one per entry in contained).
+     Worst case: All lookups succeed at maximum depth with many entries to check.
+
+     Strategy:
+     1. Generate container with worst-case BST structure (uniform, power-of-2 sizes)
+     2. Select entries FROM container (maintain subset relationship for no early exit)
+     3. Include deepest entry to force maximum BST traversal
+     4. Test multiple contained sizes to explore iteration count dimension
+
+     Result: ~1000 systematic worst-case benchmarks vs 100 random cases previously
```

Comment:
I think this is maybe a bit too much.  It takes about 2½ hours to benchmark just this function, which is much  longer than any of the other builtins.  Here's a list of the numbers of datapoints in the CSV file for the most intensively benchmarked builtins, and `valueContains` is much bigger than anything else.  Maybe we could reduce it to 15x15 or something.
```
    202 EqualsString
    225 AddInteger
    256 DivideInteger
    256 ExpModInteger
    256 MultiplyInteger
    300 EqualsByteString
    400 ConstrData
    400 EqualsData
    400 LookupCoin
    400 MkPairData
    400 SerialiseData
    441 AppendByteString
    441 AppendString
    500 ChooseList
    625 AndByteString
   1052 ValueContains
```